cmd/test.go

```diff
@@ -185,11 +200,131 @@ func opaTest(args []string) int {
 
 	txn, err := store.NewTransaction(ctx, storage.WriteParams)
 	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	runner, reporter, err := compileAndSetupTests(ctx, store, txn, modules, bundles)
+	if err != nil {
+		store.Abort(ctx, txn)
 		fmt.Fprintln(os.Stderr, err)
 		return 1
 	}
 
-	defer store.Abort(ctx, txn)
+	success := true
+	for i := 0; i < testParams.count; i++ {
+		exitCode := runTests(ctx, txn, runner, reporter)
+		if exitCode != 0 {
+			success = false
+			store.Abort(ctx, txn)
+			if testParams.watch {
+				break
+			}
+			return exitCode
+		}
+	}
+
+	if success {
+		store.Abort(ctx, txn)
+	}
+
+	if !testParams.watch {
+		return 0
+	}
+
+	done := make(chan struct{})
+	go startWatcher(ctx, args, inmem.NewWithOpts(inmem.OptRoundTripOnWrite(false)), done)
+
+	signal.Notify(testParams.stopChan, syscall.SIGINT, syscall.SIGTERM)
+
+	<-testParams.stopChan
+	done <- struct{}{}
+	return 0
+}
+
+func startWatcher(ctx context.Context, paths []string, store storage.Store, done chan struct{}) {
+	watcher, err := pathwatcher.CreatePathWatcher(paths)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Error creating path watcher: ", err)
+		os.Exit(1)
+	}
+	readWatcher(ctx, watcher, paths, store, done)
+}
+
+func readWatcher(ctx context.Context, watcher *fsnotify.Watcher, paths []string, store storage.Store, done chan struct{}) {
+	for {
+
+		fmt.Fprintln(testParams.output, strings.Repeat("*", 80))
+		fmt.Fprintln(testParams.output, "Watching for changes ...")
+		select {
+		case evt := <-watcher.Events:
+			removalMask := fsnotify.Remove | fsnotify.Rename
+			mask := fsnotify.Create | fsnotify.Write | removalMask
+			if (evt.Op & mask) != 0 {
+				removed := ""
+				if (evt.Op & removalMask) != 0 {
+					removed = evt.Name
+				}
+				processWatcherUpdate(ctx, paths, removed, store)
+			}
+		case <-done:
+			watcher.Close()
+			return
+		}
+	}
+}
+
+func processWatcherUpdate(ctx context.Context, paths []string, removed string, store storage.Store) {
+	filter := loaderFilter{
+		Ignore: testParams.ignore,
+	}
+
+	var loadResult *initload.LoadPathsResult
+
+	err := pathwatcher.ProcessWatcherUpdate(ctx, paths, removed, store, filter.Apply, testParams.bundleMode, func(ctx context.Context, txn storage.Transaction, loaded *initload.LoadPathsResult) error {
```

Comment:
This is a very long line, consider spreading the argument list over multiple lines.

Reply:
Done